### ISSUE TITLE
Window on the main monitor blinks

### ISSUE THREAD

**Issue by Groovounet** [glfw/glfw#527]
When creating multiple windows (3, each fullscreen running on a difference GPU (AMD + NVIDIA + Intel)), the main window is blinking when getting the focus again. I click on each window, on monitors 1 and 2, the window stays still, as expected but on monitor 0, it clicks as if GLFW tries to restore the window while it's not iconified.

```
#include <GLFW/glfw3.h>

void key_callback(GLFWwindow* Window, int Key, int Scancode, int Action, int Mods)
{
	if(Key == GLFW_KEY_ESCAPE && Action == GLFW_PRESS)
		glfwSetWindowShouldClose(Window, GL_TRUE);
}

void main_window(GLFWmonitor* Monitor)
{
	int Width = 0, Height = 0;
	glfwGetMonitorPhysicalSize(Monitor, &Width, &Height);

	glfwWindowHint(GLFW_RESIZABLE, GL_FALSE);
	glfwWindowHint(GLFW_VISIBLE, GL_TRUE);
	glfwWindowHint(GLFW_AUTO_ICONIFY, GL_FALSE);
	glfwWindowHint(GLFW_DECORATED, GL_FALSE);
	glfwWindowHint(GLFW_FLOATING, GL_FALSE);
	glfwWindowHint(GLFW_SRGB_CAPABLE, GL_FALSE);

	glfwWindowHint(GLFW_CLIENT_API, GLFW_OPENGL_API);
	glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4);
	glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
	glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
	glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);

#	if defined(_DEBUG)
	glfwWindowHint(GLFW_OPENGL_DEBUG_CONTEXT, GL_TRUE);
#	else
	glfwWindowHint(GLFW_OPENGL_DEBUG_CONTEXT, GL_FALSE);
#	endif

	GLFWvidmode const* Mode = glfwGetVideoMode(Monitor);
	glfwWindowHint(GLFW_RED_BITS, Mode->redBits);
	glfwWindowHint(GLFW_GREEN_BITS, Mode->greenBits);
	glfwWindowHint(GLFW_BLUE_BITS, Mode->blueBits);
	glfwWindowHint(GLFW_ALPHA_BITS, Mode->blueBits);
	glfwWindowHint(GLFW_REFRESH_RATE, Mode->refreshRate);

	GLFWwindow* Window = glfwCreateWindow(Mode->width, Mode->height, "Krolik", Monitor, nullptr);
	glfwSetWindowPos(Window, 0, 0);
	glfwSetKeyCallback(Window, key_callback);
	glfwMakeContextCurrent(Window);

	while(true)
	{
		glfwPollEvents();

		if (glfwWindowShouldClose(Window))
			break;

		glClearColor(1.0f, 0.5f, 0.0f, 1.0f);
		glClear(GL_COLOR_BUFFER_BIT);

		glfwSwapBuffers(Window);
	}

	glfwDestroyWindow(Window);
	Window = nullptr;
}

int main(int argc, char* argv[])
{
	glfwInit();

	int MonitorCount = 1;
	GLFWmonitor** Monitors = glfwGetMonitors(&MonitorCount);

	std::vector<std::thread> Threads(static_cast<std::size_t>(MonitorCount));

	for(std::size_t Index = 0; Index < Threads.size(); ++Index)
		Threads[Index] = std::thread(main_window, Monitors[Index]);

	for(std::size_t Index = 0; Index < Threads.size(); ++Index)
		Threads[Index].join();

	glfwTerminate();

	return 0;
}
```

**Comment by elmindreda** [comment]
Which GPU is on monitor zero? Is Aero enabled?

**Comment by Groovounet** [comment]
- Aero enabled yes.
- GeForce 970

**Comment by Groovounet** [comment]
Also, other monitors use different GPUs if that matters.

**Comment by elmindreda** [comment]
Window creation and event processing are only permitted in the main thread for portability reasons, so I moved those. This particular program works on Windows but it wouldn't for example have worked on OS X. Also, some calls were no-ops in the context they were called, so I removed them.

```cpp
#include <GLFW/glfw3.h>

#include <thread>
#include <vector>
#include <atomic>

std::atomic<bool> running(true);

void key_callback(GLFWwindow* Window, int Key, int Scancode, int Action, int Mods)
{
    if(Key == GLFW_KEY_ESCAPE && Action == GLFW_PRESS)
        running = false;
}

void thread_main(GLFWwindow* Window)
{
    glfwMakeContextCurrent(Window);

    while(running)
    {
        glClearColor(1.0f, 0.5f, 0.0f, 1.0f);
        glClear(GL_COLOR_BUFFER_BIT);

        glfwSwapBuffers(Window);
    }
}

GLFWwindow* create_window(GLFWmonitor* Monitor)
{
    glfwWindowHint(GLFW_RESIZABLE, GL_FALSE);
    glfwWindowHint(GLFW_AUTO_ICONIFY, GL_FALSE);
    glfwWindowHint(GLFW_SRGB_CAPABLE, GL_FALSE);

    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4);
    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
    glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
    glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);

#   if defined(_DEBUG)
    glfwWindowHint(GLFW_OPENGL_DEBUG_CONTEXT, GL_TRUE);
#   else
    glfwWindowHint(GLFW_OPENGL_DEBUG_CONTEXT, GL_FALSE);
#   endif

    GLFWvidmode const* Mode = glfwGetVideoMode(Monitor);
    glfwWindowHint(GLFW_RED_BITS, Mode->redBits);
    glfwWindowHint(GLFW_GREEN_BITS, Mode->greenBits);
    glfwWindowHint(GLFW_BLUE_BITS, Mode->blueBits);
    glfwWindowHint(GLFW_REFRESH_RATE, Mode->refreshRate);

    GLFWwindow* Window = glfwCreateWindow(Mode->width, Mode->height, "Krolik", Monitor, nullptr);
    glfwSetKeyCallback(Window, key_callback);

    return Window;
}

int main(int argc, char* argv[])
{
    glfwInit();

    int MonitorCount = 1;
    GLFWmonitor** Monitors = glfwGetMonitors(&MonitorCount);

    std::vector<std::thread> Threads(static_cast<std::size_t>(MonitorCount));

    for(std::size_t Index = 0; Index < Threads.size(); ++Index)
        Threads[Index] = std::thread(thread_main, create_window(Monitors[Index]));

    while(running)
        glfwWaitEvents();

    for(std::size_t Index = 0; Index < Threads.size(); ++Index)
        Threads[Index].join();

    glfwTerminate();

    return 0;
}
```

Back to the bug. Are you using a GLFW release, and if so, which one? If not, what commit is your snapshot at? The bug may potentially be related to #24 or the fix for it that caused #519. Looking into it and your other issue right now.

**Comment by Groovounet** [comment]
I am using version 3.1.1 so maybe #519 is guilty. 

Are you sure about  glfwWaitEvents();? 

I tried to call glfwWaitEvents in the main thread as well but it seems that by key callback never get called if I don't call glfwWaitEvents in my threaded render loop.

**Comment by Groovounet** [comment]
So actually it looks like that to work the window and the events need to be created and executed from the same thread.

**Comment by elmindreda** [comment]
> So actually it looks like that to work the window and the events need to be created and executed from the same thread.

Yes, that's the rule on Windows. On OS X, creation and processing are both tied to the main thread. On X11 they can both be done on any thread, including different ones.

**Comment by Spasi** [comment]
Happens here too. Windows 8.1, AMD GPU, dual-monitor setup on same GPU. Some observations:

1. I cannot reproduce it with multiple GLFW windows. I'm assuming it happens for @Groovounet because different GLFW windows are on different GPUs. It also doesn't happen if the GLFW window loses focus to the desktop or cmd. It does happen if focus is lost to any other application.
2. I don't think it's related to #24 or #519, because it also happens without enabling vsync.
3. It has a very noticeable effect on performance. Running a Gears demo without vsync, it starts at ~5200 fps. Losing focus to the desktop, it stays at 5200. Losing focus to another app (and the flicker happens), it drops down to ~3200 fps.

**Comment by hoinshembl** [comment]
Just a few comments to what Spasi wrote:
In my case it doesn't matter to what the window looses its focus to, switching to Desktop or cmd makes no difference.
I was able to reproduce the issue with multiple GLFW windows (one supposed to run borderless at 1920*1080 on the left (0,0) primary monitor as well as a smaller one which correctly was borderless at a smaller size on the second monitor), which made it even worse. I'm pretty sure there now were 2 flickers instead of one (but not always, though there was at least one all the time).

I'm using Windows 7, along with a NVIDA card (and no other graphics hardware - neither onboard nor integrated in the CPU).

**Comment by Spasi** [comment]
Looked into this a bit further and it seems like it's a feature of Windows. Some random references:

http://en.sfml-dev.org/forums/index.php?topic=9839.0
https://bitbucket.org/rude/love/issues/816/borderless-windows-at-native-resolution
http://stackoverflow.com/questions/22259067/opengl-non-exclusive-fullscreen-mode-a-k-a-fullscreen-borderless-window
https://code.google.com/p/chromium/issues/detail?id=20

The last one in particular is interesting:

With a non-auto-hide taskbar, using a borderless windows at (0,0) that matches the resolution of the primary display, results in the pseudo-fullscreen mode that we're seeing. The most important effect is that the taskbar is hidden, instead of showing on top of the window.

With an auto-hide taskbar, I'm seeing the same problem that affected so many Chrome users. The few taskbar pixels that are normally visible are hidden and you cannot auto-show the taskbar anymore.

**Comment by elmindreda** [comment]
@Spasi About the performance impact, see [Priority Boosts](https://msdn.microsoft.com/en-us/library/windows/desktop/ms684828.aspx).

**Comment by elmindreda** [comment]
My theory is that the flicker is DWM (un-)redirecting the full screen window. That would explain why it only happens when focus changes for the window on the primary monitor, i.e. the monitor whose ICD is being used to render. It also resembles the flicker when X11 compositors do the same.

**Comment by theagentd** [comment]
I am also getting this using two GTX 770s with 3 monitors connected. It also happens when using only 2 monitors on a single GPU without SLI. Every time a fullscreen window gets a focus change all connected monitors flicker.

**Comment by elmindreda** [comment]
If anyone has any clue what to do or even where to look, please speak up. As it is, this is too vague to be a blocking issue for 3.1.2.

**Comment by elmindreda** [comment]
I can reliably reproduce this _very_ visibly on my non-Optimus Nvidia laptop with an external monitor.

Random observations so far (on this setup): The flicker does not appear if the windows are not continually rendered to. Swap interval does not prevent the flicker either way. Each time focus shifts from one window to the other, there are two flickers. Each time focus shifts to or from a non-GLFW window, there is a single flicker.

**Comment by theagentd** [comment]
That sounds like the same thing I'm getting.

**Comment by theagentd** [comment]
If you have two monitors you can easily see this happen when the focus switches between two fullscreen windows.

**Comment by elmindreda** [comment]
The flicker does not occur if the window either has decorations or if it doesn't exactly match the monitor in size and position, strengthening my belief that this is Windows (un-)redirecting rendering directly to the monitor framebuffer. This is the desired behaviour, even if the transition could be smoother.

If anyone finds a method to avoid the flicker that doesn't involve disabling this optimization, I'll happily implement it.

**Comment by JamesDrakeUK** [comment]
I've also hit this issue (Windows 10, multi-monitor, single GTX 770 gfx) and wanted to share the workaround I found in case it helps someone else;

You can hoodwink windows into not entering fullscreen-exclusive/direct render mode on the primary monitor by building a borderless window that is one pixel taller than the actual monitor resolution eg:

```java
glfwWindowHint(GLFW_DECORATED , GL_FALSE);
window = glfwCreateWindow(width, height+1, "Hello World!", NULL, NULL);
glfwSetWindowPos(window, x, y);
```
(setup fullscreen windows as normal on the other monitors)

This will have pretty much the same behaviour as a fullscreen window but Windows wont do the horrible focus/defocus blinking. I expect there will be some performance loss associated with this, but for my application I saw no impact.

**Comment by theagentd** [comment]
I think I have an idea what is going on here. Here's a detailed timeline of what happens when I go to fullscreen from a 2560x1440 desktop to a 1920x1080 fullscreen window:

1. The screen goes black for 0.5-1.0 second as the monitor switches resolutions. Both monitors are black during this.
2. For a split second, I can see my Windows desktop running at 1920x1080.
3. Both monitors switch to black a second time for a split second.
3. The game finally goes fullscreen at 1920x1080.

My guess here is that the first flicker comes from switching resolutions and the second flicker comes from setting the fullscreen window. Is there no way of combining the two steps so that we at least don't get 2 flickers? It'd make tabbing in and out of our games faster and possibly more stable as well.

**Comment by soulthreads** [comment]
If anyone is still interested in a more "proper" sort of solution, disabling window flag `WS_POPUP` fixes this problem. I don't exactly know why though, but this is a general problem with fullscreen OpenGL applications on Windows.

**Comment by jaredmulconry** [comment]
This is an issue I've ran into recently while trying to get 3 borderless fullscreen windows across my 3 monitors. If focus switches between any of them, the primary window flickers a couple of times.
This occurs using the latest version of GLFW on my Windows 10 system with an Nvidia GTX660Ti graphics card.

**Comment by ghuser404** [comment]
@elmindreda, are we still looking for a proper fix or removing `WS_POPUP` would be enough? This issue has been bugging me for a while now. If it matters, I know at least one AAA game that doesn't use `WS_POPUP` in borderless fullscreen mode.

**Comment by elmindreda** [comment]
@ghuser404 The flicker still occurs for me without `WS_POPUP`, with or without combinations of `WS_SYSMENU` and `WS_MINIMIZEBOX`, both with and without DWM/Aero.

**Comment by ghuser404** [comment]
@elmindreda, could you share code so that I could test it on my machine?

**Comment by ghuser404** [comment]
@elmindreda, to be more clear, I had it working with my modifications to GLFW, so I was wondering what you modified and how you used it, that was still giving you this behaviour, and see if I would also get it with your code.

**Comment by elmindreda** [comment]
@ghuser404 Yup, seems it works with Nvidia drivers but not Intel ones. That matches what appeared to be the case when inspecting this with API Monitor; that every ICD implements the exclusive fullscreen hack separately.

I'll get you code (if you still want it?) I'm just running a lot of threads finishing up 3.3 at the moment.

**Comment by ghuser404** [comment]
@elmindreda, yeah, code would be awesome, just to be on par. I could test it on different GPUs then.

**Comment by gpugfx** [comment]
Re flashing with full screen window, I've hit this before, and fixed one of the conditions that causes it to happen (see below) in a non-GLFW-based app.  But I'm looking into another weird case that causes it to happen for some users now.  That prompted me to revisit this thread and see if anyone had come up with the full fix.

First, here's what I know.  In my case (and the cases I've fixed for users), this flashing only happens:

1. with Full-Screen Exclusive (FSE) mode compatible windows, 
2. which are not single buffered (e.g. double buffered),
3. which have focus,
4. when on startup the app both: 1) changes the screen resolution and 2) creates the Full-Screen Exclusive window, and
5. when the app is actively calling SwapBuffers() on the Full-Screen window.

Remove any of these from the loop and it doesn't happen (or goes away, in the case of removing window focus).  It's the act of calling SwapBuffers when the window has focus that triggers a flash.

FSE requires res == screen res, WS_POPUP (and no additional flags which precludes FSE), and for the window to have focus.  FSE gets you a perf boost and other benefits such as synchronizing with scan-out (on Win7) as your frame takes a faster path through the display system.  However, switching into and out of FSE mode (on Win7 at least) causes a full-screen flash.  

Here's what I suspect.  I started by theorizing that the flashes might be changes into and/or out of FSE mode.  That seems a reasonable guess since switching focus to/from a FSE window generates the same flash, and a FSE-compatible window is involved.  It struck me as odd that SwapBuffers on a single fullscreen window would be causing changes into (and/or out) FSE mode.  Why would that happen?  Well, what are we doing?  With double-buffering, under-the-hood we're switching swap-chain images.  What if those images in the swap chain aren't being created such that they are all FSE-compatible for some reason?  For instance, it could be that some of the images are being created with the "old" res and some with the "new" (fullscreen) res.  Or something like that.  That could theoretically happen if the new screen/monitor resolution isn't fully active, applied, and propagated "before" the window (and its underlying swap chain+images) is created.

With that insight, I fixed one scenario which triggers this full-screen flashing case.  That being where the full-screen window was created "before" the screen resolution was changed.  The flashing caused by that was pretty easy to reproduce.  The solution was just to flip these and change the screen resolution before creating the full-screen window.  That fix has been sufficient for quite a while now.

However, now I'm getting reports from a few users that they're seeing the flashing thing again with full-screen windows even with the above fix, but I haven't been able to reproduce this with the same GPU+GPU driver+app+app config.  So I'm left suspecting that this might be a race condition.  Perhaps the screen resolution change isn't fully complete, active, and propagated before the window (and its underlying swap chain) is created.  To explore that theory, I'm going to see if I can find and call the equivalent of XSync() or glFinish() for the windows display system (possibly in combination with a sleep if I have to) after the res change to try and ensure the res change is fully complete before creating the window.  Maybe GdiFlush(), but this sounds more like an XFlush() / glFlush() which is not what I need.  I want to wait until the res change is fully complete (in GDI/DWM/WDDM/GPU graphics driver/...whatever).

Sorry for the book, but I'm hoping this triggers some ideas from someone else to continue further so we can fix this annoying Windows problem which has been plaguing folks for years.  I'll check back.

**Comment by ghuser404** [comment]
@gpugfx, this happens to me when I'm just creating window with the size of the monitor native resolution - I believe that is different from changing display's resolution.

**Comment by elmindreda** [comment]
I've also been able to consistently trigger this without changing the resolution.

The requirements have seemed to be only a focused, top-most, undecorated window whose size and position match a monitor viewport and that you swap the buffers of a context using that window as a draw target.

**Comment by ghuser404** [comment]
Yep, but top-most is optional too, it seems.

**Comment by gpugfx** [comment]
In case it might be useful to someone here...

I ended up solving my Full-Screen Exclusive (FSE) screen-flashing-on-swap problem (on Windows, in a non-GLFW-based app) by switching from changing monitor resolution/bpp/refresh rate using a method based on:

- EnumDisplaySettings()
- ChangeDisplaySettings()

to one based on:

- for all monitors/screens we want to change the resolution/bpp/refresh rate of:
--   EnumDisplaySettings()
--   ChangeDisplaySettingsEx( ... ( CDS_UPDATEREGISTRY | CDS_NORESET ), 0 )
- ChangeDisplaySettingsEx( 0,0,0,0,0 )

For details on the latter, see the code snippet on this page: [ChangeDisplaySettingsEx](https://msdn.microsoft.com/en-us/library/windows/desktop/dd183413(v=vs.85).aspx)

**Comment by codedcosmos** [comment]
I am also getting this issue on the plasma desktop.

**Comment by W4RH4WK** [comment]
Any news regarding this issue?

**Comment by leecommamichael** [comment]
What worked for me:

I was using `GetPrimaryMonitor` and then passing that to `CreateWindow`. I just passed null instead and nothing blinked.

**Comment by alex-tkachenko** [comment]
I have the similar issue. Using Cocos2d-x 4.0, Win32. The game works fine in fullscreen on every system I tried (even on Virtual Box). But on MacBook Pro late 2013  (GeForce 750M, Windows 10 with Boot Camp) I'm getting constant black blinking in fullscreen and windowed fullscreen. Tried latest GLFW build from the master - without success. Does anyone know any workaround or fix?
